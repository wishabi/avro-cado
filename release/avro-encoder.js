"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// import Avro = require("avsc");
const Avro = require("avsc");
const util_1 = require("./util");
const config_1 = require("./config");
const rp = require("request-promise");
/**
 * Register the specified schema for the specified topic under the
 * specifed subject. The subject used is: <topic>-<key|value>
 *
 * @param topic - the topic for which to register a schema
 * @param schemaRegistry - the schema registry host
 * @param numRetries - the number of retries to register before throwing an error
 * @param schema - the schema to register
 * @param suffix - the key or value portion of the subject
 *
 * @return - A Promise holding the id of the schema in the schema registry
 */
const registerSchema = ({ subject, schemaRegistry, numRetries, schema }) => __awaiter(this, void 0, void 0, function* () {
    // null schema so nothing to register
    if (!schema) {
        return -1;
    }
    // craft the REST call to the schema registry
    const req = {
        method: "POST",
        uri: `${schemaRegistry}/subjects/${subject}/versions`,
        headers: [{ Accept: config_1.ACCEPT_HEADERS }],
        body: {
            schema: JSON.stringify(schema)
        },
        json: true,
        resolveWithFullResponse: true
    };
    let schemaId;
    let error = null;
    // implement retry on certain status/reason codes
    for (let i = 0; i <= numRetries; i += 1) {
        try {
            schemaId = (yield rp(req)).body.id;
            // make sure to clear any previous errors
            error = null;
            break;
        }
        catch (err) {
            // save the error
            error = err;
            // see if it can be retried
            const retry = util_1.handleError(err, config_1.encodeLogger, "register schema");
            if (retry && i + 1 <= numRetries) {
                // try and try again until we run out of retries
                continue;
            }
            // fatal error
            break;
        }
    }
    if (error) {
        config_1.encodeLogger.error(`Registering schema for ${subject}  failed`, error, {
            subject,
            schema
        });
        throw new Error(`Failed to register schema for subject ${subject}`);
    }
    config_1.encodeLogger.info(`Registered schema for subject ${subject}`, {
        schemaId,
        subject
    });
    return schemaId;
});
/**
 * Encode an Avro value into a message, as expected by Confluent's Kafka Avro
 * deserializer.
 *
 * @param message - message to encode
 */
const encodeData = (payload, schemaId, schema, length = 1024) => {
    if (payload === null) {
        // no payload so return it
        return null;
    }
    if (schemaId === -1) {
        // Bufferify the payload
        return Buffer.from(JSON.stringify(payload));
    }
    const buf = new Buffer(length);
    buf[0] = 0; // magic byte
    buf.writeInt32BE(schemaId, 1);
    const pos = schema.encode(payload, buf, 5);
    if (pos < 0) {
        // the buffer was too short, we need to resize
        return encodeData(payload, schemaId, schema, length - pos);
    }
    return buf.slice(0, pos);
};
const genMessageEncoder = ({ schema, schemaId }) => (message) => {
    return encodeData(message, schemaId, schema);
};
const createAvroSchema = ({ wrapUnions }, schema) => {
    if (schema) {
        return Avro.Type.forSchema(schema, { wrapUnions });
    }
    return null;
};
/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/
exports.createTopicEncoder = ({ schemaRegistry, numRetries, wrapUnions, subject, schema }) => __awaiter(this, void 0, void 0, function* () {
    // Aggregare the configuration values with defaults
    const mergedOptions = util_1.aggregateOptions(config_1.optionsDefault, {
        schemaRegistry,
        numRetries,
        wrapUnions,
        subject,
        schema
    });
    // topic info
    const topicInfo = {
        schema: null,
        schemaId: -1
    };
    // register the value and key schemas and parse them
    [topicInfo.schema, topicInfo.schemaId] = yield Promise.all([
        createAvroSchema(mergedOptions, schema),
        registerSchema(mergedOptions)
    ]);
    return genMessageEncoder(topicInfo);
});
//# sourceMappingURL=avro-encoder.js.map