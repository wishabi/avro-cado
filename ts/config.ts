import { Options } from "../ts/types/types";
import { Logger } from "@flipp/flipp-node-logger";

export const ACCEPT_HEADERS: string =
  "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";

export const optionsDefault: Options = {
  schemaRegistry: "http://localhost:8081",
  numRetries: 10,
  wrapUnions: "auto",
  subject: "DEFAULT",
  schema: null
};

export const encodeLogger: Logger = new Logger("encodeLogger");
export const decodeLogger: Logger = new Logger("decodeLogger");
