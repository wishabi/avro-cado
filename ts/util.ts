import { Options } from "./types/types";

export const RETRY_STATUS_CODE_500 = 500;
export const RETRY_ERROR_CODE_50002 = 50002;
export const RETRY_ERROR_CODE_50003 = 50003;

/**
 * Decide whether a schema registry error is eligible to
 * be retried
 *
 * @param err - the error in question
 * @return - true iff the error is eligible for retry
 */
export const handleError = (err: any): boolean => {
  if (
    err.statusCode &&
    err.statusCode === RETRY_STATUS_CODE_500 &&
    err.error &&
    err.error.error_code &&
    (err.error.error_code === RETRY_ERROR_CODE_50003 ||
      err.error.error_code === RETRY_ERROR_CODE_50002)
  ) {
    return true;
  }

  return false;
};

/**
 * Merge the user specified configuration options with the
 * default ones
 *
 * @param defaultConf - the default configuration options
 * @param overrideConf - the user specified configuration options
 * @return - merged configuration options object
 */
export const aggregateOptions = (
  defaultConf: Options,
  overrideConf: Options
): Options => {
  const aggOptions = Object.assign({}, defaultConf, overrideConf);

  if (!overrideConf) {
    return aggOptions;
  }

  // determine how to handle unions
  const validWrapOptions: Array<string> = ["always", "never", "auto"];
  if (
    validWrapOptions.filter(opt => opt === aggOptions.wrapUnions).length === 0
  ) {
    aggOptions.wrapUnions = defaultConf.wrapUnions;
  }

  return aggOptions;
};
