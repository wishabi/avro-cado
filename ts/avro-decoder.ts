import * as Avro from "avsc";
import * as rp from "request-promise";
import { ACCEPT_HEADERS, processOptions } from "./config";
import {
  DecodeFunc,
  DecoderInfo,
  Options,
  SchemaResolverFunc
} from "./types/types";
import { aggregateOptions, handleError } from "./util";

/**
 * Retrieve the Avro schema from the schema registry. On an error
 * eligible for retry, try up to the configure number of times
 *
 * @param subject - the subject under which the schema is registered
 * @param schemaRegistry - the schema registry path
 * @param numRetries - the number of retries
 * @param id - the schema id that should be retrieved
 * @return - The Avro schema requested
 */
export const retrieveSchema = async (
  { subject, schemaRegistry, numRetries }: Options,
  id: number
) => {
  const req = {
    uri: `${schemaRegistry}/schemas/ids/${id}`,
    headers: { Accept: ACCEPT_HEADERS },
    json: true,
    resolveWithFullResponse: true
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return JSON.parse((await rp(req)).body.schema);
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to retrieve schema for subject ${subject} with id ${id} :: ${
      error.message
    }`
  );
};

/**
 * Retrieve the Avro schema from the schema registry, parse it,
 * and create a resolver to the local Avro schema.
 *
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param schema - the target Avro schema for the resolver
 * @return - The Avro schema resolver used to convert messages
 *           encoded with the specified Avro schema to the local
 *           Avro schema format
 */
export const genCreateSchemaResolver = (
  opts: Options
): SchemaResolverFunc => async (
  id: number,
  schema: Avro.Type
): Promise<Avro.Resolver> => {
  /**
   * Retrieve the schema by ID from the schema registry ...
   */
  const msgSchema = await retrieveSchema(opts, id);

  /**
   * ... parse the schema ...
   */
  try {
    const avSourceSchema = Avro.Type.forSchema(msgSchema);

    /**
     * ... and create a resolver to the schema we know how to consume
     */
    return schema.createResolver(avSourceSchema);
  } catch (err) {
    throw new Error(
      `Local schema is not compatible with schema from registry with id ${id} :: ${
        err.message
      }`
    );
  }
};

/**
 * Create an Avro decoder based on the specified parameters and
 * return a closure that takes an encoded Buffer and decodes it
 *
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param createSchemaResolver - a function used to create the resolver
 *                               object that will be used to decode the payload
 * @param resolversMap - a hashmap keys on the schema ids where the value is a
 *                       promise holding a schema resolver
 * @param buffer - The Avro encoded buffer
 * @return - The Avro decoded object that conforms to the specified schema
 */
export const genPayloadDecoder = ({
  schema,
  createSchemaResolver,
  resolversMap
}: DecoderInfo): DecodeFunc => async (
  buffer: Buffer | null
): Promise<object> => {
  // Do not attempt to avro decode null value
  if (!buffer) {
    return buffer;
  }

  /**
   * Extract the schema id with which the buffer was encoded
   */
  const id: number = buffer.slice(1, 5).readInt32BE(0);

  if (!resolversMap[id]) {
    /**
     * We do not have a resolver from the source schema to the
     * schema we support. Go and get a resolver.
     *
     * IMPORTANT
     *
     * We store a promise of a resolver in our hashmap. The reason is that
     * if we have multiple messages being handled asynchronously, we do not
     * want all of them to flood the schema registry with REST calls
     * to retrieve the schema
     */
    resolversMap[id] = createSchemaResolver(id, schema);
  }

  return schema.decode(buffer.slice(5), 0, await resolversMap[id]).value;
};

/*
 *****************************************************************
 *                      EXPORTED INTERFACE                       *
 *****************************************************************
 */

export const createDecoder = (opts: Options): DecodeFunc => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  // decoder info
  const decoderInfo: DecoderInfo = {
    subject: mergedOpts.subject,
    schema: Avro.Type.forSchema(mergedOpts.schema),
    resolversMap: {},
    createSchemaResolver: null
  };

  decoderInfo.createSchemaResolver = genCreateSchemaResolver(mergedOpts);

  return genPayloadDecoder(decoderInfo);
};
