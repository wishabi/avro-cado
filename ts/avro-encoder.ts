import * as Avro from "avsc";
import axios, { AxiosResponse } from "axios";
import { ACCEPT_HEADERS, processOptions } from "./config";
import { EncodeFunc, Options } from "./types/types";
import { handleError } from "./util";

/**
 * Register the specified schema for the specified topic under the
 * specified subject.
 *
 * On an error eligible for retry, keep trying up to the specified number
 * of times. Once all retries have been exhausted, any error is considered
 * a fatal one.
 *
 *
 * @param subject - the subject under which to register the schema
 * @param schemaRegistry - the schema registry host
 * @param numRetries - the number of retries to register before throwing an error
 * @param schema - the schema to register
 *
 * @return - A Promise holding the id of the schema in the schema registry
 */
export const registerSchema = async ({
  subject,
  schemaRegistry,
  schema,
  numRetries,
}: Options): Promise<number> => {
  // craft the REST call to the schema registry
  const req = {
    method: "POST",
    url: `${schemaRegistry}/subjects/${subject}/versions`,
    headers: { Accept: ACCEPT_HEADERS },
    data: {
      schema: JSON.stringify(schema),
    },
    json: true,
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return ((await axios(req)) as AxiosResponse).data.id;
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to register schema for subject ${subject} :: ${error.message}`,
  );
};

/**
 * Create a function that takes a message JSON object
 * and Avro encodes it
 *
 * @param payload - the JSON object to Avro encode
 * @return - The Avro encoded object as a Buffer
 */
export const genMessageEncoder = (
  schema: Avro.Type,
  schemaId: number,
): EncodeFunc => (payload: object): Buffer => {
  if (payload === null) {
    // no payload so return it
    return null;
  }

  // create the avro header
  const header = Buffer.alloc(5);

  // magic byte
  header.writeInt8(0, 0);

  // schema id
  header.writeInt32BE(schemaId, 1);

  // Avro encode the payload
  const buffer: Buffer = schema.toBuffer(payload);

  return Buffer.concat([header, buffer]);
};

/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/

export const createEncoder = async (opts: Options): Promise<EncodeFunc> => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  return genMessageEncoder(
    Avro.Type.forSchema(mergedOpts.schema, {
      wrapUnions: mergedOpts.wrapUnions,
    }),
    await registerSchema(mergedOpts),
  );
};
