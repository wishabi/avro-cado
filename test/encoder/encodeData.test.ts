jest.unmock("avsc");

import { encodeData } from "../../ts/avro-encoder";
import * as Avro from "avsc";

const AVRO_SCHEMA = {
  type: "record",
  name: "TestMessage",
  namespace: "com.flipp.node.kafka.TestMessage",
  doc: "Properties related to a TestMessage.",
  fields: [
    {
      name: "key",
      type: "string",
      doc: "The the key for the message"
    },
    {
      name: "text",
      type: "string",
      doc: "The text for the message"
    }
  ]
};

const AVRO_SCHEMA_OBJ = Avro.Type.forSchema(AVRO_SCHEMA);

const AVRO_SCHEMA_ID: number = 1;

const TEST_MESSAGE = {
  key: `1`,
  text: `A simple test message 1`
};

const TEST_MESSAGE_INVALID = {
  noKey: `1`,
  text: `A simple test message 1`
};

describe("encodeData", () => {
  it("should AVRO encode a valid message", () => {
    expect.assertions(1);

    const encoded: Buffer = encodeData(
      TEST_MESSAGE,
      AVRO_SCHEMA_ID,
      AVRO_SCHEMA_OBJ
    );
    expect(encoded).toMatchSnapshot();
  });

  it("should NOT AVRO encode a valid message with no schema", () => {
    expect.assertions(1);

    const encoded: Buffer = encodeData(TEST_MESSAGE, -1, null);
    expect(encoded).toMatchSnapshot();
  });

  it("should NOT AVRO encode a null", () => {
    expect.assertions(1);

    const encoded: Buffer = encodeData(null, -1, null);
    expect(encoded).toMatchSnapshot();
  });

  it("should AVRO encode a null", () => {
    expect.assertions(1);

    const encoded: Buffer = encodeData(null, AVRO_SCHEMA_ID, AVRO_SCHEMA_OBJ);
    expect(encoded).toMatchSnapshot();
  });

  it("should raise exception when payload does NOT match schema", () => {
    expect.assertions(1);

    expect(() =>
      encodeData(TEST_MESSAGE_INVALID, AVRO_SCHEMA_ID, AVRO_SCHEMA_OBJ)
    ).toThrowError();
  });
});
