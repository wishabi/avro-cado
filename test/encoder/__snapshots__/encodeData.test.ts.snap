// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encodeData should AVRO encode a null 1`] = `null`;

exports[`encodeData should AVRO encode a valid message 1`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    1,
    2,
    49,
    46,
    65,
    32,
    115,
    105,
    109,
    112,
    108,
    101,
    32,
    116,
    101,
    115,
    116,
    32,
    109,
    101,
    115,
    115,
    97,
    103,
    101,
    32,
    49,
  ],
  "type": "Buffer",
}
`;

exports[`encodeData should NOT AVRO encode a null 1`] = `null`;

exports[`encodeData should NOT AVRO encode a valid message with no schema 1`] = `
Object {
  "data": Array [
    123,
    34,
    107,
    101,
    121,
    34,
    58,
    34,
    49,
    34,
    44,
    34,
    116,
    101,
    120,
    116,
    34,
    58,
    34,
    65,
    32,
    115,
    105,
    109,
    112,
    108,
    101,
    32,
    116,
    101,
    115,
    116,
    32,
    109,
    101,
    115,
    115,
    97,
    103,
    101,
    32,
    49,
    34,
    125,
  ],
  "type": "Buffer",
}
`;
